source/particle/world.cc

```diff
@@ -617,6 +614,158 @@ namespace aspect
 
     namespace internal
     {
+
+      /**
+       * Wrapper around dealii::FEPointEvaluation to choose number of components dynamically.
+       */
+      template <int dim>
+      class DynamicFEPointEvaluation
+      {
+        public:
+          DynamicFEPointEvaluation(const unsigned int first_component, const unsigned int n_components)
+            : first_component (first_component),
+              n_components(n_components)
+          {}
+
+          virtual ~DynamicFEPointEvaluation() = default;
+
+          int first_component;
+          int n_components;
+
+          virtual void evaluate(const ArrayView<double> &solution_values,
+                                const EvaluationFlags::EvaluationFlags flags) = 0;
+
+          virtual
+          std::vector<Tensor<1,dim>>
+          get_gradient(const unsigned int evaluation_point) const = 0;
+
+          virtual
+          std::vector<double>
+          get_value(const unsigned int evaluation_point) const = 0;
+
+      };
+
+
+
+      template <int dim>
+      Tensor<1,dim> to_tensor(const Tensor<1,dim> &in)
+      {
+        return in;
+      }
+
+
+
+      template <int dim>
+      Tensor<1,1> to_tensor(const double in)
+      {
+        Tensor<1,1> result;
+        result[0] = in;
+        return result;
+      }
+
+
+
+      template <int dim, int n_components>
+      Tensor<1,n_components,Tensor<1,dim>> to_tensor2(const Tensor<1,n_components,Tensor<1,dim>> &in)
+      {
+        return in;
+      }
+
+
+
+      template <int dim, int n_components>
+      Tensor<1,1,Tensor<1,dim>> to_tensor2(const Tensor<1,dim> &in)
+      {
+        Tensor<1,1,Tensor<1,dim>> result;
+        result[0] = in;
+        return result;
+      }
+
+
+
+      template <int dim, int n_components>
+      class DynamicFEPointEvaluationImpl: public DynamicFEPointEvaluation<dim>
+      {
+        public:
+          DynamicFEPointEvaluationImpl(NonMatching::MappingInfo<dim> &mapping,
+                                       const FiniteElement<dim> &fe,
+                                       const unsigned int        first_selected_component)
+            : DynamicFEPointEvaluation<dim>(first_selected_component, n_components),
```

Comment:
What do you do if the `n_components` are not consecutive? FEPointEvaluation only assumes consecutive components, but as far as I remember we now allow merging the base elements of non-consecutive components?

Reply:
This only works for consecutive items (and this is what introspection returns). The only place where we do magic with non-consecutive items is with matrix block locations.